### PR TITLE
Add apiserver metric for response sizes

### PR DESCRIPTION
**Pull request by shyamjvs**
Fixes https://github.com/kubernetes/kubernetes/issues/47728

This should help us understand GET/LIST call latencies better. It'll also help catch differences in object sizes across kubemark and real cluster.
I'm labelling the metrics by namespace (hoping that there won't be toooo many of them).

/cc @smarterclayton @gmarek 
cc @kubernetes/sig-scalability-misc @kubernetes/sig-api-machinery-misc @lavalamp @wojtek-t

### CODE REVIEW

**Comment by shyamjvs** [comment]
In an offline discussion today @wojtek-t suggested that we categorize the metric as namespaced/non-namespaced (instead of labelling by namespace name) to keep the no. of metrics bounded.
I agree with it and changed it so.

**Comment by shyamjvs** [comment]
@wojtek-t BTW - during the discussion you asked how useful this is except for testing. I think it would be good for cluster admins to:
- know how many 'list all' calls are being made in their cluster: Too many => not very good (as @smarterclayton pointed in https://github.com/kubernetes/kubernetes/issues/47728#issuecomment-309582688)
- know how big responses requests are generating in general: This can be a good signal to catch presence of malicious components (that intentionally solicit huge responses)
- know how sizes of objects in general are changing over time: E.g endpoints, configmaps, etc

Overall this can potentially help improve cluster security and understand object/list sizes in etcd.
Moreover you propose SLI for measuring list call latency based on #objects returned in your [apimachinery SLIs doc](https://docs.google.com/document/d/1Q5qxdeBPgTTIXZxdsFILg7kgqWhvOwY8uROEf0j5YBw/edit?ts=596f5ee0#). I think a similar extension of our metrics might be required to measure that.

**Comment by wojtek-t** [comment]
I'm still not convinced that this PR is what we want. In particular some answers to your comments:

> know how many 'list all' calls are being made in their cluster: Too many => not very good (as @smarterclayton pointed in #47728 (comment))

For that it is enough to have request count (which we already have)

> know how big responses requests are generating in general: This can be a good signal to catch presence of malicious components (that intentionally solicit huge responses)

Then maybe we should report only very large responses (e.g. metric should be "count of requests with responses larger than say 1MB" (maybe splitted by client or sth)).

> know how sizes of objects in general are changing over time: E.g endpoints, configmaps, etc

If that's the purpose, let's expose this as a metric.

**Comment by shyamjvs** [comment]
I mostly agree with your opinion. However:

> For that it is enough to have request count (which we already have)

If we decide to not have the response-size metric, then I can remove it and add extra label (for namespace) to our request_count metric. But if we are having the metric, then I don't think it's a good idea to add namespace to request count for following reason:
- We're already unbounded due to client (as you pointed out), adding new label there would simply double them
- We can find what kinds of abusive calls are being made from response_size metric (for e.g <LIST,endpoints> calls are having huge response sizes), so we can simply use the other metric (which is request_count) to find out which client is making maximum such calls. This will help us avoid having 'client' label for both metrics.

> Then maybe we should report only very large responses (e.g. metric should be "count of requests with responses larger than say 1MB" (maybe splitted by client or sth)).

I'm fine with making this change. However this probably just means trading off extra information for saving some prometheus metric.Observe() calls. I'm not a prometheus expert to comment about the cost of those calls, so I'm neutral both ways. (cc @piosz for any prometheus-related suggestions)

> If that's the purpose, let's expose this as a metric.

IMO what is more useful than monitoring sizes of objects in etcd (some of which are probably unused and just lie there), is to measure sizes of objects/lists that actually flow through the system and how those change over time. And measuring response sizes (more or less) captures that. Sorry I didn't make this distinction clear earlier.

**Comment by piosz** [comment]
@brancz @matthiasr @fabxc for Prometheus related questions.

**Comment by brancz** [comment]
From a general Prometheus perspective the implementation looks ok. I do share the same thoughts on the usefulness of these metrics as @wojtek-t, summaries are relatively expensive on the client so we should consider whether this is worth the cost.

**Comment by shyamjvs** [comment]
@brancz Thanks for the lead. To clarify, are you suggesting not having the metric at all or saving Observe() calls on the metric by recording only high values?

**Comment by brancz** [comment]
I was arguing on having them at all, but the cost of summaries is naturally lower for less values being added, as every added value triggers calculations of the overall summary. In contrast to histograms the values you get through summaries basically already have to be what you want to query, meaning most aggregations are useless on them, which is why I'm questioning the usefulness of these metrics.

> know how sizes of objects in general are changing over time: E.g endpoints, configmaps, etc

This one for example is not really something I'm seeing any Prometheus user do. Typically people have relatively sort retention periods making the "change over time" only visible within that retention period (15 days by default). This might be different in other monitoring solutions, but that's the perspective from Prometheus.

> know how big responses requests are generating in general: This can be a good signal to catch presence of malicious components (that intentionally solicit huge responses)

Do we have a concrete case for this? (iirc etcd values are limited to 1Mb, in which case I don't see how that can be possible in the first place)

**Comment by shyamjvs** [comment]
Thanks for the valuable info.

> I was arguing on having them at all, but the cost of summaries is naturally lower for less values being added, as every added value triggers calculations of the overall summary.

Makes sense. I'm fine with changing it to recording only high response sizes (we can iterate on that based on performance). But I'm slightly hesitant to remove the metric totally due to the benefits I mentioned. That said, if people agree (probably based on real user/admin usecases) that we don't really need this, I'd gladly pull it back.

> In contrast to histograms the values you get through summaries basically already have to be what you want to query, meaning most aggregations are useless on them, which is why I'm questioning the usefulness of these metrics.

IIUC we don't need any fancy aggregation here except the standard 50th, 90th and 99th percentiles (just like for e.g response_latencies metric) which summaries are enough to provide.

> This one for example is not really something I'm seeing any Prometheus user do.

I think there's a confusion here (if not, please clarify) about measuring sizes of objects stored in etcd vs sizes of responses returned by the server. I was suggesting the latter instead of the former (in https://github.com/kubernetes/kubernetes/pull/49117#issuecomment-316702136). And measuring response sizes seems to be a very common use-case of prometheus users (as mentioned in the [prometheus reference](https://prometheus.io/docs/concepts/metric_types/#histogram))

> Do we have a concrete case for this? (iirc etcd values are limited to 1Mb, in which case I don't see how that can be possible in the first place)

The intention here is to measure response sizes from apiserver rather than that from etcd.
Btw - by 1Mb do you mean limit of object size in etcd (as that still means we can have responses bigger than that due to lists) or responses from etcd (which might be true due to pagination, but I lack enough knowledge here)?

**Comment by smarterclayton** [comment]
The 1mb limit doesn't apply anymore - you could easily fetch 500mb from etcd is a single call, and the distribution of latency and sizes is absolutely important to know how large your db is growing.

**Comment by smarterclayton** [comment]
We removed "namespace" from the other apiserver metrics because of cardinality (we have clusters with 10k namespaces).

**Comment by smarterclayton** [comment]
When we evaluate compression and other paging improvements coming to both etcd and the apiserver this will be relevant to track how that changes the calls.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 1]
```diff
@@ -62,14 +62,13 @@ var (
 	responseSizesSummary = prometheus.NewSummaryVec(
 		prometheus.SummaryOpts{
 			Name: "apiserver_response_sizes_summary",
-			Help: "Response size summary in bytes for each verb, resource, subresource and namespace.",
+			Help: "Response size summary in bytes for each verb, resource, subresource and if it's namespaced.",
 			// Make the sliding window of 1h.
 			MaxAge: time.Hour,
 		},
-		[]string{"verb", "resource", "subresource", "namespace"},
+		[]string{"verb", "resource", "subresource", "namespaced"},
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: We refer to this as a "scope" and we have two of them - "namespace" and "cluster".  We should use those variables and values for this.

**Comment by mml** [comment]
I don't quite understand why we've chosen summary over histogram for the type here.

**Comment by shyamjvs** [comment]
@mml Quoting from [prometheus reference](https://prometheus.io/docs/practices/histograms/):

```
The essential difference between summaries and histograms is that summaries calculate streaming φ-quantiles on the client side and expose them directly, while histograms expose bucketed observation counts and the calculation of quantiles from the buckets of a histogram happens on the server side using the histogram_quantile() function.
```

So:
- Firstly, if we're just looking for just quantiles and don't need other fancy aggregations, summary is enough
- Assuming in future we define SLIs/SLOs on these quantiles (which usually is the case), it is better to use summary over histogram because accuracy of quantiles would be much higher in the former (as the latter uses buckets there's some error, unless we create really fine buckets)
- Third and the most important reason IMO is using histograms => defining the right buckets for response sizes, which is very challenging to do for different <verb, resource> pairs. For e.g endpoints objects are really big whereas namespace objects are small

Regarding performance - histograms are server-side intensive while summaries are client-side intensive, but both are on the master so this shouldn't matter too much.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 1 above)
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Changed it everywhere except in installer.go where "scope" is already being used by restmapper's scope.

**Comment by shyamjvs** [comment]
@smarterclayton @mml Addressed comments. PTAL.

**Comment by matthiasr** [comment]
On Thu, Jul 20, 2017, 22:04 Shyam JVS <notifications@github.com> wrote:

> I'm fine with changing it to recording only high response sizes (we can
> iterate on that based on performance).
>
But what is the meaning of a metric that only records part of what is
happening? As an operator, how am I supposed to make sense of that?



> IIUC we don't need any fancy aggregation here except the standard 50th,
> 90th and 99th percentiles (just like for e.g response_latencies metric)
> which summaries are enough to provide.
>

The problem is not that you can't do other aggregations, the problem is
that you *cannot aggregate quantiles* once they are calculated. There is
simply not enough information. This makes them completely meaningless for
anyone operating an HA apiserver setup (we do), and fundamentally prevents
answering questions like "is the 99th percentile across this customer's GKE
higher than 100MB".

Summaries are *useless* outside of a very narrow scope, and if you are
willing to not even include all events they make no sense within that
scope. If all my responses but one are smaller than the magical threshold,
then that single response is the 99th, 90th and 50th percentile of… of
itself, really, because that's all you will be able to say at that point.

On the other hand, if you include all events, most responses are small,
that one 500MB blob is not going to move the needle on the aggregated
percentile

It seems to me the question you are trying to answer is "how many
abnormally large responses are there" for a few values of abnormal. That is
*exactly* the data that a histogram collects, at negligible realtime cost
(because it only increments a handful of numbers).


To summarize, please use histograms. Summaries are not the right tool for
various reasons, most of them boiling down to "except in toy situations,
cluster operators won't be able to answer the questions you are adding
these metrics to answer." The performance issues of recording them are just
the frosting on top.

>

**Comment by smarterclayton** [comment]
Go ahead and squash, lgtm

**Comment by smarterclayton** [comment]
Barring any further discussion from mml, will give him time to do that

**Comment by shyamjvs** [comment]
@smarterclayton Thanks for looking at it. Squashed them. Will address the other comments.

**Comment by shyamjvs** [comment]
@matthiasr Thanks for the explanation. That's a useful perspective from cluster-ops side and looks reasonable. Histograms have more information than summaries in general, provided the buckets are chosen well. Do you have some suggestions for what the buckets should look like? Considering:

- IMO we shouldn't have too many of them as that gets multiplied by the cardinality of <verb, resource>
- having too few of them doesn't give good accuracy for percentiles (which is fundamental to the use of buckets)
- we should reasonably capture distributions of different object kinds (for e.g namespaces/services are in bytes-KBs while endpoints/configmaps can range in bytes - 100s of MBs, also list response vs get responses)

> It seems to me the question you are trying to answer is "how many abnormally large responses are there" for a few values of abnormal.

To clarify, that is just one of the reasons like I mentioned above (probably interesting from a security PoV). As a cluster user you might also want to see what ranges of sizes generally are seen for different calls, how many of them are namespaced, etc to know the kind of load the apiserver is subject to. You might see many large responses, but that might be ok as there probably were many many many small responses and they compare well - so having just large response-sizes might leave you clueless here.

**Comment by shyamjvs** [comment]
I've changed the metric to histogram with exponential buckets from 1KB to 1GB with a step factor of 10. That looks reasonable for now and we probably can fine tune it later if required.

@mml Could you PTAL now?

**Comment by matthiasr** [comment]
Powers of ten is a good approach :+1:

**Comment by mml** [comment]
Thanks, @shyamjvs.  What @matthiasr said is largely what was behind my question.  You can answer a lot more questions with histograms than summaries, and we should probably only add summaries in the case where we specifically need an exact measurement of a specific quantile or we need to save space.

In fact, we should document that somewhere.  Do we have a doc on our monitoring/instrumentation philosophy?

/lgtm

**Comment by shyamjvs** [comment]
I couldn't find any k8s-specific doc for this, but the prometheus doc on best-practices has all this nicely explained - https://prometheus.io/docs/practices/histograms

**Comment by shyamjvs** [comment]
Squashed the commits. Applying lgtm based on https://github.com/kubernetes/kubernetes/pull/49117#issuecomment-317126594

**Comment by shyamjvs** [comment]
@smarterclayton Also any idea why swagger-spec needed a change due to this (I can't see anything obvious)?

**Comment by smarterclayton** [comment]
Removing labels until we know why swagger changed

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 2]
```diff
@@ -554,17 +554,17 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 			versionedObject = defaultVersionedObject
 		}
 		reqScope.Namer = action.Namer
-		namespaced := ""
+		namespaced := "Cluster"
```
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: @smarterclayton It's because of this change I made (in order to be consistent with our naming scheme you suggested).

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 3]
```diff
@@ -554,17 +554,17 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
...
 		if apiResource.Namespaced {
-			namespaced = "Namespaced"
+			namespaced = "Namespace"
```
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: And this too

**Comment by shyamjvs** [comment]
@smarterclayton Are you ok with the above changes that caused swagger update?

**Comment by brancz** [comment]
@mml such a document exist :slightly_smiling_face: https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md ... it links out to the general Prometheus suggestions, but it might still be worth it to drop a summarizing paragraph in there as the Prometheus docs go into a lot of detail (not a bad thing, just for a short point of reference it would be good to have a paragraph of our own).

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go (hunk 2 above)
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: This change isn't necessary

Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: This just affects swagger, it shouldn't be part of the metrics change.  We already have a constant and type for "rest scoping", so we shouldn't be keying off a string here for the metrics instrumentation, we should be keying off the value of apiResource.Namespaced and then passing that to metrics, leaving the swagger alone

Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: Cool. I'll change it back. Are you also fine with the scope label for the metric taking values "Namespace" and "" then?

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go (hunk 3 above)
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: @smarterclayton Does this change seem fine or you want me to put this back too?

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go (hunk 2 above)
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: There was a race with my last response, sorry. Yes I'll make the change accordingly. Thanks.

**Comment by shyamjvs** [comment]
@smarterclayton Fixed swagger. Could you PTAL finally?

**Comment by shyamjvs** [comment]
There seems to be a slight confusion in naming b/w non-namespaced vs all-namespaced requests. While RESTScope-wise, non-namespaced resources ("root") are differentiated from all-namespaced resources ("namespace"), but they are similar wrt swagger spec (i.e. both are labelled by "" for namespace). For the purpose of our metric, the latter scheme seems more relevant. WDYT?

**Comment by smarterclayton** [comment]
Sorry, will try to describe better.  Has been a while since we were in this.

Our fundamental API concept is a "rest scope" which has either "root" or "namespace".  Root really means "cluster" to an end user.  I think we should show "cluster" or empty string in metrics (empty string is pretty consistent with other behavior).  A call to "get a particular resource from all namespaces" is effectively the same as a call to "a cluster scoped resource".  I think for user purposes, the scope argument for "get from all namespaces" and "get a resource that does not live in a namespace" should be the same.  I think we can reconcile this as saying some calls have a "scope", which is "namespace" (namespace scoping).  All calls not in a namespace scope today are the empty string scope "" (which is then consistent).

We shouldn't change swagger since it can break users, so its terminology is at least consistent.

So I'd expect to see:

```
var scope string
if restScope.ScopeName() == "namespace" {
  scope = "namespace"
}
```
and in prometheus
```
foo_metric{scope=""}
foo_metric{scope="namespace"}
```

**Comment by smarterclayton** [comment]
Unless that causes a problem / violates a rule in prometheus.  In which case, we should use `scope="cluster"`

**Comment by shyamjvs** [comment]
Thanks a lot for the explanation. That was perfectly in line what I was thinking too i.e all-namespaced and non-namespaced resources both should be seen similarly for the purpose of our metric.
If user wants to distinguish the cases he just needs to look at the resource in question.

**Comment by shyamjvs** [comment]
However a small correction wrt the following you said:

```
So I'd expect to see:

var scope string
if restScope.ScopeName() == "namespace" {
  scope = "namespace"
}
```

This will make scope equal to "namespace" for all-namespace requests too (just because the resource concerned is namespaced). So I guess we don't want to do this, but instead just reuse the variable we calculate for the purpose of swagger.

**Comment by shyamjvs** [comment]
@smarterclayton Fixed on comments (including moving ResponseWriterDelegator to an interface). PTAL now.

**Comment by smarterclayton** [comment]
Yes, correct.  We only want to show a "namespace" scoped query as `scope="namespace"` int prometheus

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go [hunk 4]
```diff
@@ -109,7 +109,11 @@ func WithMaxInFlightLimit(
 						}
 					}
 				}
-				metrics.MonitorRequest(r, strings.ToUpper(requestInfo.Verb), requestInfo.Resource, requestInfo.Subresource, "", errors.StatusTooManyRequests, time.Now())
+				scope := ""
+				if requestInfo.Namespace != "" {
+					scope = "Namespaced"
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: Probably should be `namespace`

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go [hunk 5]
```diff
@@ -60,7 +60,11 @@ func WithTimeoutForNonLongRunningRequests(handler http.Handler, requestContextMa
 		}
 		now := time.Now()
 		metricFn := func() {
-			metrics.MonitorRequest(req, strings.ToUpper(requestInfo.Verb), requestInfo.Resource, requestInfo.Subresource, "", http.StatusInternalServerError, now)
+			scope := ""
+			if requestInfo.Namespace != "" {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: Probably should be `namespace`

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go (hunk 4 above)
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go (hunk 5 above)
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: Done.

**Comment by matthiasr** [comment]
In Prometheus, setting a label to an empty string is equivalent to not
setting it at all. It kinda works most of the time but it's discouraged
because it can cause strange effects in aggregations.

**Comment by shyamjvs** [comment]
@matthiasr Changed from empty string to "cluster" instead as Clayton suggested. Thanks for pointing this out.

**Comment by shyamjvs** [comment]
Closing and reopening to rekick tests.

**Comment by shyamjvs** [comment]
@kubernetes/test-infra-maintainers The tests aren't kicking. Any idea why?

**Comment by krzyzacy** [comment]
PR Jenkins seems down cc @rmmh

**Comment by shyamjvs** [comment]
@smarterclayton Fixed on comments. Could you PTAL?
